File: src/main/java/gargoyle/rpycg/service/ScriptConverter.java
package gargoyle.rpycg.service;

import gargoyle.rpycg.ex.AppUserException;
import gargoyle.rpycg.ex.MalformedScriptException;
import gargoyle.rpycg.fx.FXContextFactory;
import gargoyle.rpycg.model.ModelItem;
import gargoyle.rpycg.model.ModelType;
import gargoyle.rpycg.model.VarType;
import gargoyle.rpycg.util.Check;

import java.text.MessageFormat;
import java.util.*;

public final class ScriptConverter {
    private static final String LC_ERROR_FAIL_TYPE = "error.fail-type";
    private static final String LC_ERROR_VALUE_TYPE = "error.value-type";
    private static final String LC_ERROR_WRONG_MODEL_TYPE = "error.wrong-model-type";
    private static final String LC_ERROR_WRONG_TYPE = "error.wrong-type";
    private final ResourceBundle resources;

    public ScriptConverter() {
        resources = FXContextFactory.currentContext().loadResources(ScriptConverter.class)
                .orElseThrow(() ->
                        new AppUserException(AppUserException.LC_ERROR_NO_RESOURCES, ScriptConverter.class.getName()));
    }

    public ModelItem fromScript(final Iterable<String> lines) {
        final ModelItem root = ModelItem.createMenu("", "");
        ModelItem menu = root;
        for (final String raw : lines) {
            if (raw.isBlank()) {
                continue;
            }
            final String line = raw.trim();
            if ('<' == line.charAt(0)) {
                final String substring = line.substring(1);
                final int posName = substring.lastIndexOf(';');
                final String label;
                final String name;
                if (0 < posName) {
                    label = substring.substring(0, posName).trim();
                    name = substring.substring(posName + 1).trim();
                } else {
                    label = substring;
                    name = substring;
                }
                final ModelItem child = ModelItem.createMenu(label, name);
                menu.addChild(child);
                menu = child;
                continue;
            }
            if ('>' == line.charAt(0)) {
                menu = menu.getParent();
                if (null == menu) {
                    return root;
                }
                continue;
            }
            String expr = line;
            final int posLabel = expr.lastIndexOf(';');
            final String label;
            if (0 < posLabel) {
                label = expr.substring(posLabel + 1).trim();
                expr = expr.substring(0, posLabel).trim();
            } else {
                label = "";
            }
            VarType type = null;
            if (!expr.isEmpty() && ')' == expr.charAt(expr.length() - 1)) {
                final int open = expr.lastIndexOf('(');
                if (0 < open) {
                    final String typeValue = expr.substring(open + 1, expr.length() - 1).trim();
                    try {
                        type = VarType.valueOf(typeValue.toUpperCase(Locale.ENGLISH));
                    } catch (final IllegalArgumentException e) {
                        throw new MalformedScriptException(
                                MessageFormat.format(resources.getString(LC_ERROR_WRONG_TYPE),
                                        typeValue, Arrays.toString(VarType.values())), e);
                    }
                    expr = expr.substring(0, open).trim();
                }
            }
            final String name;
            final String value;
            final int posEq = expr.lastIndexOf('=');
            if (0 < posEq) {
                final String val = expr.substring(posEq + 1).trim();
                name = expr.substring(0, posEq).trim();
                final int len = val.length();
                final char first = val.charAt(0);
                final char last = val.charAt(len - 1);
                if ('\'' == first && '\'' == last || '\"' == first && '\"' == last) {
                    type = VarType.STR;
                    value = val.substring(1, len - 2).trim();
                } else {
                    value = val;
                }
            } else {
                name = expr;
                value = "";
            }
            if (null == type) {
                if (value.isBlank()) {
                    type = VarType.STR;
                } else if (Check.isFloat(value)) {
                    type = VarType.FLOAT;
                } else if (Check.isInteger(value)) {
                    type = VarType.INT;
                } else {
                    throw new MalformedScriptException(resources.getString(LC_ERROR_FAIL_TYPE));
                }
            }
            if (!value.isBlank()) {
                switch (type) {
                    case INT:
                        if (!Check.isInteger(value)) {
                            throw new MalformedScriptException(resources.getString(LC_ERROR_VALUE_TYPE));
                        }
                        break;
                    case FLOAT:
                        if (!Check.isFloat(value)) {
                            throw new MalformedScriptException(resources.getString(LC_ERROR_VALUE_TYPE));
                        }
                        break;
                    case STR:
                        break;
                }
            }
            menu.addChild(ModelItem.createVariable(type, label, name, value));
        }
        return root;
    }

    public List<String> toScript(final ModelItem item) {
        final String label = item.getLabel();
        final String name = item.getName();
        final ModelType modelType = item.getModelType();
        switch (modelType) {
            case MENU:
                final List<String> lines = new LinkedList<>();
                if (!name.isBlank()) {
                    lines.add(MessageFormat.format("<{0};{1}", label, name));
                }
                for (final ModelItem child : item.getChildren()) {
                    lines.addAll(toScript(child));
                }
                if (!name.isBlank()) {
                    lines.add(">");
                }
                return lines;
            case VARIABLE:
                final String value = item.getValue();
                final String keyword = null == item.getType() ? "" : item.getType().getKeyword();
                return Collections.singletonList(MessageFormat.format("{0}{1}({2}){3}",
                        name,
                        value.isBlank() ? "" : '=' + value,
                        keyword,
                        label.isBlank() || label.equals(name) ? "" : ';' + label
                ));
            default:
                throw new IllegalStateException(resources.getString(LC_ERROR_WRONG_MODEL_TYPE) + " " + modelType);
        }
    }
}

File: app/src/main/kotlin/gargoyle/rpycg/service/ScriptConverter.kt
package gargoyle.rpycg.service

import gargoyle.fx.FxContext
import gargoyle.fx.FxUtil.get
import gargoyle.rpycg.ex.MalformedScriptException
import gargoyle.rpycg.model.ModelItem
import gargoyle.rpycg.model.ModelType
import gargoyle.rpycg.model.VarType
import java.text.MessageFormat
import java.util.ResourceBundle

class ScriptConverter {
    private val resources: ResourceBundle =
        FxContext.current.loadResources(ScriptConverter::class) ?: error("No resources {ScriptConverter}")

    fun fromScript(lines: List<String>): ModelItem {
        val root = ModelItem.createMenu("", "")
        var menu: ModelItem = root
        for (raw in lines) {
            if (raw.isBlank()) continue
            val line = raw.trim()
            if (line.startsWith('<')) {
                ModelItem.createMenu(
                    line.trimStart('<').substringBefore(';').trim(),
                    line.trimStart('<').substringAfter(';').trim()
                ).also {
                    menu.addChild(it)
                    menu = it
                }
                continue
            }
            if (line.startsWith('>')) {
                if (menu.parent == null) return root
                menu = menu.parent!!
                continue
            }
            val expr0: String = if (line.contains(';')) line.substringBefore(';').trim() else line
            val label: String = if (line.contains(';')) line.substringAfter(';').trim() else ""
            var type: VarType? = null
            val expr1: String
            if (expr0.endsWith(')') && expr0.contains('(')) {
                expr1 = expr0.substringBefore('(').trim()
                type = try {
                    VarType.find(expr0.substringAfter('(').substringBefore(')').trim())
                } catch (e: NoSuchElementException) {
                    throw MalformedScriptException(
                        MessageFormat.format(
                            resources[LC_ERROR_WRONG_TYPE],
                            e.message,
                            VarType.values().contentToString()
                        ), e
                    )
                }
            } else {
                expr1 = expr0
            }
            val name: String
            val value: String
            if (expr1.contains('=')) {
                val value0 = expr1.substringAfterLast('=').trim()
                name = expr1.substringBeforeLast('=').trim()
                if ((value0.startsWith('\'') && value0.endsWith('\'')) ||
                    (value0.startsWith('\"') && value0.endsWith('\"'))
                ) {
                    type = VarType.STR
                    value = value0.trim('\'', '\"').trim()
                } else {
                    value = value0
                }
            } else {
                name = expr1
                value = ""
            }
            if (null == type) type = when {
                value.isBlank() -> VarType.STR
                null != value.toFloatOrNull() -> VarType.FLOAT
                null != value.toIntOrNull() -> VarType.INT
                else -> throw MalformedScriptException(resources[LC_ERROR_FAIL_TYPE])
            }
            if (value.isNotBlank())
                when (type) {
                    VarType.INT -> if (null == value.toIntOrNull())
                        throw MalformedScriptException(resources[LC_ERROR_VALUE_TYPE])
                    VarType.FLOAT -> if (null == value.toFloatOrNull())
                        throw MalformedScriptException(resources[LC_ERROR_VALUE_TYPE])
                    VarType.STR -> {}
                }
            menu.addChild(ModelItem.createVariable(type, label, name, value))
        }
        return root
    }

    fun toScript(item: ModelItem): List<String> =
        when (item.modelType) {
            ModelType.MENU ->
                if (item.name.isNotBlank()) listOf("<${item.label};${item.name}") +
                        item.getChildren().flatMap { toScript(it) } + listOf(">")
                else item.getChildren().flatMap { toScript(it) }
            ModelType.VARIABLE ->
                listOf("${item.name}${if (item.value.isBlank()) "" else "=${item.value}"}(${item.type.keyword})${if (item.label.isBlank() || item.label == item.name) "" else ";${item.label}"}")
        }

    companion object {
        private const val LC_ERROR_FAIL_TYPE = "error.fail-type"
        private const val LC_ERROR_VALUE_TYPE = "error.value-type"
        private const val LC_ERROR_WRONG_MODEL_TYPE = "error.wrong-model-type"
        private const val LC_ERROR_WRONG_TYPE = "error.wrong-type"
    }
}
